antipackage.py

```diff
@@ -1,130 +1,401 @@
+        Parameters
+        ----------
+        pathlist : list
+            Of the form `[github, username, repository]`
+            and indicates the repository that should be
+            accessed
+        name : str
+            Should be either 'sha', 'tag' or 'branch' and
+            indicates the version access method for the repo
+        value : str
+            specific to `name` and indicated the specific
+            repository version being referenced"""
+        self.name = name
+        self.value = value
+        self.url = self._url(pathlist)
+        self._git_init()
+
+    def _url(self, pathlist):
+        """Format the GitHub url"""
+        url = self.host['top'].format(*pathlist)
+        url += self.host[self.name].format(self.value)
+        return url
+
+    def _git_init(self):
+        """Initialize data from GitHub"""
+        temp_file = self.fetch()
+        with open(temp_file, 'r') as f:
+            data = json.load(f)
+            self.sha = self._sha(data)
+
+    def fetch(self):
+        """Retrieve data from GitHub"""
+        temp_file, response = urlretrieve(self.url)
+        if not response['Status'] == '200 OK':
+            e = 'at '+self.url
             try:
-                self._install_module(fullname)
-            except InstallError:
-                if os.path.isfile(self._install_path(fullname)):
-                    print('Using existing version: ', fullname)
-                else:
-                    print('Error installing/updating module: ', fullname)
+                with open(temp_file, 'r') as f:
+                    out = json.load(f)
+                e = "message "+e+" = '"+out['message']+"'"
+            except:
+                pass
+            raise GitHubError(e)
+        return temp_file
+
+    def _sha(self, data):
+        """Return the sha"""
+        if self.name!='sha':
+            if self.name == 'tag':
+                data = self._handle_tag(data)
+            return data['commit']['sha']
+        else:
+            return self.value
+
+    def _handle_tag(self, data):
+        """Retrieve data from a """
+        for tag in data:
+            if tag['name']==self.value:
+                return tag
+        raise RetrieveError("No tag with name: '%s'" % value)
+
+# - - - - - - - - - - - -
+# Administrative Importer
+# - - - - - - - - - - - -
+
+class Importer(object):
+
+    github = GitHubHook
+
+    def __init__(self):
+        sys.path.append(_BASE_DIR)
 
     def find_module(self, fullname, path=None):
-        # print('find_module', fullname, path)
-        if fullname.startswith('github'):
-            self._make_package(fullname)
-        return None
+        fullname = repr_conv(fullname)
+        prefix = fullname.split('.')[0]
+        if hasattr(self, prefix):
+            source = getattr(self, prefix)(fullname)
+            if self.require(source):
+                source.install()
+                source.required = True
+
+    def require(self, source):
```

Comment:
All of this logic should be put into a single method on the `GitHubHook`. Maybe we want a single top level method on `GitHubHook` called `install_or_update`?